src/sentry/models/authenticator.py

```diff
@@ -340,9 +341,32 @@ def _set_secret(self, secret):
     def make_otp(self):
         return TOTP(self.secret)
 
+    def _get_otp_counter_cache_key(self, counter):
+        if self.authenticator is not None:
+            return 'used-otp-counters:%s:%s' % (
+                self.authenticator.user.id,
+                counter,
+            )
+
+    def check_otp_counter(self, counter):
+        cache_key = self._get_otp_counter_cache_key(counter)
+        return cache_key is None or cache.get(cache_key) != '1'
```

Comment:
_Technically_ this is an `O(n)` network requests for the cache. Probably fine though since I assume `window` won't be large.